### ISSUE TITLE
Only save non-default input variables to household

### ISSUE THREAD

**Issue by MaxGhenis** [PolicyEngine/policyengine-api#230]
For example, in #200 I linked [this household](https://policyengine.org/us/household?focus=householdOutput.netIncome&household=6874), which I defined by only setting the state to NY; all other variables were default. But it now breaks because it referred to `co_tanf_countable_income`, which is no longer in the system. Can we tie household IDs only to non-default variables, so we only break households with variables that changed?

**Comment by anth-volk** [comment]
From testing, if I access the same linked household in the app at the moment, I receive an empty white box where the household output should be, but if I create a new household with all of the same details, I do receive accurate output. I'll dig deeper, as well, but I would imagine fixing this involves the app, core, and api packages?

**Comment by anth-volk** [comment]
Just wanted to provide some documentation on the problem, so as to better formulate a response. The problem seems to run like this: 
1. On lines 117-120 of `policyengine-app/src/pages/HouseholdPage.jsx`, the HouseholdPage component makes an API request to fetch a household under a policy, using the user's input household data
2. This is in the form of a GET request to `/<country_id>/household/<household_id>/policy/<policy_id>`, with the policy_id being the country current law if no reform has been provided by the user
3. This triggers the `get_household_under_policy` API controller, defined on line 105 of `policyengine-api/policyengine_api/endpoints/household.py`
4. After executing some data fetching, the controller attempts to calculate household impacts in the try-except block on lines 189-203 in the same file
5. The `calculate` method utilized is defined on line 270 of `policyengine-api/policyengine_api/country.py`, which does nothing with the out-of-date household data until passing it to the relevant country package's `Simulation` class as its `situation` parameter on line 294 of the same file, which is itself is defined in `policyengine-core/policyengine_core/simulations/simulation.py` on line 35
6. The API `calculate` method then enqueues calculations for each variable in the household data file, before passing them to core's `Simulation` class's own `calculate` method, which then passes that to its own `_calculate` method (defined on line 436 of `policyengine-core/policyengine_core/simulations/simulation.py`)
7. It looks like this is the point where the error occurs, as on line 449 of the same file, the method raises a ValueError if the calculation variable (in this case, co_tanf_countable_income) in the household data object does not exist in the tax/benefit system.

The household data object is itself created via the following process:
1. Inside of an effect hook that begins on line 59 of `policyengine-app/src/pages/HouseholdPage.jsx`, the app creates a new household via the `createDefaultHousehold` function
2. As defined on line 59 of `policyengine-app/src/api/variables.js`, `createDefaultHousehold` defines a series of household defaults, then on line 131, it invokes the `addYearlyVariables` function to add tax/benefit variables to the situation
3. This is in turn defined on line 31 of the same file, and I think this may be the first root of the problem – this function iterates over the entire variable metadata of a country situation and adds these variables, including default tax policies, to the default household before returning a constructed object

I wrote all of that out because it seems like fixing the problem would likely require cross-package changes:
* `policyengine-app` would likely need changes to `addYearlyVariables`, perhaps even its elimination (or merely the addition of a few empty object layers that would then contain any changes to defaults later in the application’s lifecycle)
* The `_calculate` method of the `Simulation` class may need altering; in particular, I’m just not sure what impact removing null calculations from the method would have
* An alternative to rewriting `addYearlyVariables` would be to do one of the following:
* When a household is fetched at step 3 of the calculation process at the top of this comment, it does contain an API version number. Perhaps the application stack (I’m not sure where this would be at the moment) could archive the country packages that correspond with the API version (or even add a country package version number), then that particular package version could be loaded by the app and core packages before doing calculations, though this would of course freeze any calculation in time and require the creation of new households each time the country package is updated
* If null values do factor into the `Simulation` class’s calculations, perhaps there could be a method to essentially “inflate” them at simulation time by comparing the household’s variables against the country metadata’s current law?

Then again, this may not be all that important; navigating to a directory with an older `household_id` creates this problem, but when I recreated all of the details exactly from the original problem, which created a new household, everything was in order.

Curious to hear what you both think @MaxGhenis @nikhilwoodruff

**Comment by nikhilwoodruff** [comment]
Thanks for the detailed write-up @anth-volk! So I actually think we can do it with a change to the API repo and the app repo, without needing to change core. The calculation function in the API repo simply looks for the null variables and fills them in: we currently add the null variables on the app side, but we could just add them in the API repo, just before passing to the calculation function as normal.

**Comment by MaxGhenis** [comment]
Is this why https://policyengine.org/us/household?focus=householdOutput.netIncome&household=32076 is broken?

**Comment by nikhilwoodruff** [comment]
Yes- [here's](https://api.policyengine.org/us/household/32076/policy/2) the API call that fails, with its error. Looks like a Maine income tax variable that was removed

**Comment by MaxGhenis** [comment]
Any update here? I just renamed `nj_child_tax_credit` to `nj_ctc` for alignment with other states and it broke all households. See also https://github.com/PolicyEngine/policyengine-us/issues/2719 which would rename a number of variables.

**Comment by MaxGhenis** [comment]
Never mind, looks like https://github.com/PolicyEngine/policyengine-api/issues/617 was the culprit with the above issue.

**Comment by nikhilwoodruff** [comment]
To summarise, closing this issue needs:
- [ ] The PolicyEngine client to create new households with *only* the variables specifically entered by the user, and no others.
- [ ] The PolicyEngine API /calculate endpoint to extend these smaller households with the full output variable set (previously the variables which had `null` values on creation and calculate as before.
- [ ] The PolicyEngine client to sanitise legacy household IDs as they are navigated to by removing unspecified user inputs and variables that no longer exist for every new time that a globally unseen (household ID, country API version) is encountered. If a user has specified a value explicitly for a variable that no longer exists, a modal should appear and the household should be deleted and the user redirected to the equivalent page without the household ID.

**Comment by MaxGhenis** [comment]
Can we bulk sanitize the household database too by clearing values that are either zero and or calculated?

**Comment by anth-volk** [comment]
This should've been closed via [PolicyEngine/policyengine-app#717](https://github.com/PolicyEngine/policyengine-app/pull/717)
